### PR TITLE
added base class

### LINKED ISSUE #23: Add a base class for the algorithms
Inherit form ABC

### PR DESCRIPTION
**Pull request by akagam1**

### CODE REVIEW

File: jeta/base.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+from abc import ABC
+from typing import Callable
+
+import jax.numpy as jnp
+from jax import random
+
+
+class JetaBase(ABC):
```
Comment (veds12) [inline jeta/base.py]:
```suggestion
class BaseLearner(ABC):
```

File: jeta/base.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        self.losses = []
+        self.seed = kwargs["seed"] if "seed" in kwargs else None
+        self.first_order = (
+            kwargs["first_order"] if "first_order" in kwargs else None
+        )  # for MAML and FOMAML
+        if self.seed is not None:
+            self.key = random.PRNGKey(self.seed)
+
+    def train(self):
+        pass
```
Comment (veds12) [inline jeta/base.py]:
```suggestion
        raise NotImplementedError
```

File: jeta/base.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    def loss(self, theta, batch):
+        pass
```
Comment (veds12) [inline jeta/base.py]:
```suggestion
        raise NotImplementedError
```

File: jeta/base.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    def test(self):
+        pass
```
Comment (veds12) [inline jeta/base.py]:
```suggestion
        raise NotImplementedError
```

File: jeta/base.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        self.fas = fas
+        self.losses = []
+        self.seed = kwargs["seed"] if "seed" in kwargs else None
+        self.first_order = (
+            kwargs["first_order"] if "first_order" in kwargs else None
+        )  # for MAML and FOMAML
+        if self.seed is not None:
+            self.key = random.PRNGKey(self.seed)
+
+    def train(self):
```
Comment (veds12) [inline jeta/base.py]: Also need to add docstrings here

**Review by veds12 — CHANGES_REQUESTED** [review]

**Comment by veds12** [comment]
@akagam1 thanks for the PR! I have left some comments. Also, do run the pre-commit before pushing so that lint issues are resolved

**Comment by abhi-glitchhg** [comment]
@akagam1 
I think you can delete the  .DS_Store files in this PR,
.DS_Store need not be tracked by git. You can add the below line in the  `.gitignore` file so that git will ignore all such files in the future. 

```bash
**/.DS_Store
```

File: jeta/base.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
+from abc import ABC
+
+from jax import random
+
+
+class BaseLearner(ABC):
+    """Base Learner class
+
+    Attributes:
```
Comment (abhi-glitchhg) [inline jeta/base.py]:
```suggestion

    Args:
```

In other docstrings, we are using `Args` only.   So to be consistent with the documentation, I would suggest using `Args` only.

Comment (abhi-glitchhg) [reply jeta/base.py]: Or if this is intentional or any docstring format is decided, then let me know.

File: jeta/base.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        self.seed = kwargs["seed"] if "seed" in kwargs else None
+        self.first_order = kwargs["first_order"] if "first_order" in kwargs else None
+        if self.seed is not None:
+            self.key = random.PRNGKey(self.seed)
+
+    def train(self, params, batch):
+        """train algorithm
+
+        Args:
+            params(ndarray): initial parameters for the model
```
Comment (abhi-glitchhg) [inline jeta/base.py]: I am not sure if the type for params will be ndarray.

File: jeta/base.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    def __init__(
+        self, batch_size: int = 32, alpha: float = 0.001, fas: int = 1, **kwargs
+    ):
+
+        self.batch_size = batch_size
+        self.alpha = alpha
+        self.fas = fas
+        self.losses = []
+        self.seed = kwargs["seed"] if "seed" in kwargs else None
+        self.first_order = kwargs["first_order"] if "first_order" in kwargs else None
```
Comment (veds12) [inline jeta/base.py]:
```suggestion

```

This isn't a general parameter for most of the M2L algorithms. So not needed in the base class

**Comment by veds12** [comment]
Additionally, @akagam1 the DS store file still shows up in your commit

File: jeta/base.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+class BaseLearner(ABC):
+    """Base Learner class
+
+    Args:
+        batch_size(int): batch size for training the algorithm on
+        alpha(float): Learning rate
+        fas(int): Fast adaptation steps
+        seed(int): integer value for generating key for random generators
+        first_order(bool): True/False, for extending MAML to FO-MAML
```
Comment (abhi-glitchhg) [inline jeta/base.py]:
```suggestion
        seed(int): integer value for generating key for random generators

```

File: jeta/base.py (hunk 7 above)
Comment (abhi-glitchhg) [reply jeta/base.py]: I think the type of the params should be a FrozenDict and not ndarray.  or am I missing anything? 

cc: @veds12

Comment (veds12) [reply jeta/base.py]: I agree @abhi-glitchhg. 

@akagam1 please fix this

**Comment by veds12** [comment]
@akagam1 please make the final changes. I'll merge after that

File: jeta/base.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+    def train(self, params, batch):
+        """train algorithm
+
+        Args:
+            params(FrozenDict): initial parameters for the model
+            batch(ndarray): training batch for the algorithm
+
+        Return:
+            new_params(ndarray): updated parameters after 1 epoch
```
Comment (abhi-glitchhg) [inline jeta/base.py]:
```suggestion
        Return:
            new_params(FrozenDict): updated parameters after 1 epoch
```

**Review by abhi-glitchhg — APPROVED** [review]
Thanks, @akagam1; changes look good to me; 

The datatypes for other arguments might need to be modified in the future, but for now, I'm okay with those.

**Review by veds12 — APPROVED** [review]

**Pull request merged**